lib/recursive-methods.rb

```diff
@@ -1,49 +1,89 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(nlogn)
+# Space complexity: O(nlogn)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError, "Error!!"
+  end
+
+  if n == 0 || n == 1
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return s
+  else
+    return s[-1] + reverse(s[0...-1])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
```

Comment:
This is not in place since you create a new string. 

Think about making a helper which takes the string, a left index and right index.  The base case is when `left_index >= right_index`  Each recursive call you swap the characters at `left_index` or `right_index` and then make a recursive call with the string, and `left_index + 1` and `right_index -1`.